Add `show-open-prs-of-forks` feature

loilo opened the pull request:
This PR adds the new `show-open-prs-of-forks` feature. It triggers in all your repositories which are forks. Because of the user-centric nature of this feature, I unfortunately can not provide test links.

### Repo Headline Hints

This feature adds a hint in the secondary repo headline, linking to your open PRs in the original repo:

#### Single Pull Request

<img src="https://user-images.githubusercontent.com/1922624/76399951-ce381600-637f-11ea-85d0-ac26f7a90200.png" alt="Head area of the loilo/mautic repo with the secondary headline stating &quot;forked from mautic/mautic with one open pull request&quot; with the &quot;one open pull request&quot; part as a link" width="420" height="120">

> If there is just one open pull request, the link will directly point there (for this screenshot: mautic/mautic#8106).

#### Multiple Pull Requests

<img src="https://user-images.githubusercontent.com/1922624/76399723-684b8e80-637f-11ea-8c99-a9f0a73ee763.png" alt="Head area of the loilo/pull-request-target repo with the secondary headline stating &quot;forked from loilo-archive/pull-request-target with 2 open pull requests&quot; with the &quot;2 open pull requests&quot; part as a link" width="420" height="120">

> If there are multiple open pull requests, the link will point to a list of them (for this screenshot: [loilo-archive/pull-request-target/pulls?q=is:pr+is:open+sort:updated-desc+author:loilo](https://github.com/loilo-archive/pull-request-target/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3Aloilo)).

### Repo Deletion Warnings

The feature also adds a paragraph to the repo deletion dialog, raising your awareness that there are unmerged outgoing PRs from this repo which you will abandon. (I ran a quick test for this — PRs are not actually deleted when the fork is gone, they're just attributed to an "unknown repo", so the warning will be about the loss of write access to the PR.)

<img src="https://user-images.githubusercontent.com/1922624/76398603-726c8d80-637d-11ea-95ca-fdf493302195.png" alt="Confirmation dialog of repository deletion with added paragraph warning about the open pull requests that will be abandoned when the repo is deleted" width="448" height="376">

Rules for link targets are the same as in the headline hints layed out above.

---

Closes #2268

jerone commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+features.add({
+	id: __featureName__,
+	description: 'Add a warning about open outgoing PRs when you\'re about to delete a repo.',
+	screenshot: 'https://user-images.githubusercontent.com/1922624/76398603-726c8d80-637d-11ea-95ca-fdf493302195.png',
```
A second screenshot won't be recognized in the settings.

loilo replied:
Good to know... but does that mean it should be removed?

fregante replied:
Set description and screenshot to an empty string. They’re not parsed at all, we just need to make typescript happy

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+		<a href={getUserPullRequestsURL(forkedRepo, user)}>your {pullRequestsCount} open pull requests</a>;
+
+	deleteDialogParagraph.after(
+		<p>
```
I considered that but ultimately rejected the thought because this already is a confirmation dialog after a button click and already has a warning that literally states that you definitely should read everything, but I guess this is as realistic as reading terms and conditions. 😁

So yeah, I agree.

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+
+	if (!forkedRepo) {
+		return;
+	}
```
Agree on the need of caching. What would be the benefit of using `forkedRepo` as a parameter instead of just using memoization on `findForkedRepo`?

fregante replied:
I think it makes sense to have functions as “pure” as possible, i.e. not read from global state. `cache.function` itself is a memoization function, so its first parameter will be used as key of the cache.

Example here: https://github.com/sindresorhus/refined-github/pull/2868/files#diff-5664f505bec7f184c9588cc203936104R90

A future update to my cache module would simplify this code even more and it would make more sense to have parameters

loilo replied:
I generally agree about the purity aspect of functions, but I don't think it's too realistic in this extension as basically all information comes from the environment. I could pass the `forkedRepo` as a parameter, but I'd still rely on `isRepoWithAccess()` and `getUsername()`.

But the example you linked has the same problem. What was the reasoning there for using `latestTag` as a parameter, but not the other data?

fregante replied:
> but I'd still rely on isRepoWithAccess() and getUsername().

Not isRepoWithAccess, that check should be outside. 

getUsername is basically a constant for the whole session, the cache doesn’t depend on it. 

> But the example you linked has the same problem. What was the reasoning there for using latestTag as a parameter, but not the other data?

latestTag is the cache key, the other data isn’t (or is already part of it). Improvements may always be possible, but we should look at the lowest common denominator.

loilo replied:
Sure, was just trying to dig deeper to gain some insight into prior art decision making. 👍

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+		search(type: ISSUE, query: "repo:${forkedRepo} is:pr is:open author:${user}", first: 1) {
+			issueCount
+			edges {
+				node {
```
Good to know. 👍

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	return search;
+}
```
Sounds reasonable.

loilo replied:
What would you consider to be an appropriate cache duration for this?

fregante replied:
Maybe half a day: `maxAge: 1 / 2`

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+		return forkSourceElement.pathname.slice(1);
+	}
+
+	return undefined;
```
Definitely. This function is actually literally copied over from the `forked-to` feature. Would feel better to put it into some utility file. Hadn't found one before, but `utils.ts` seems to be what I was looking for, so I'll probably move it over there.

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,139 @@
+}
+
+function getUserPullRequestsURL(forkedRepo: string, user: string): string {
+	return `https://github.com/${forkedRepo}/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3A${user}`;
```
Do I need to consider anything around this URL regarding GitHub Enterprise support? Would `location.origin` be more appropriate here than hardcoding the github.com domain?

busches replied:
Yes

loilo commented:
I think I've addressed all points from the previous review, except one thing: I could *not* move the `isRepoWithAccess()` check outside the function because it always returned `false` in that case (probably because it relies on a DOM check).

Also, besides the suggested simplification of the `findForkedRepo` function, I moved it to `utils.ts` (since it is also used in the `forked-to` feature) and renamed it to `getForkedRepo` (to be more in line with naming of other utility functions).

Also, my [question about GHE](https://github.com/sindresorhus/refined-github/pull/2880#discussion_r392469779) still needs to be sorted out.

fregante commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,121 @@
+}, {
+	maxAge: 1 / 2, // Stale after 12 hours
+	staleWhileRevalidate: 2,
+	cacheKey: ([forkedRepo]): string => `${__featureName__}:pr-data:${forkedRepo ?? '-'}`
```
```suggestion
	cacheKey: ([forkedRepo]): string => __featureName__ + forkedRepo
```

fregante commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,121 @@
+	deleteDialogParagraph.after(
+		<p className="flash flash-warn">
+			It will also abandon {pullRequestsLink} in <strong>{forkedRepo}</strong> and
+			you&apos;ll no longer be able to edit {pullRequestsData.count === 1 ? 'it' : 'them'}.
```
```suggestion
			you’ll no longer be able to edit {pullRequestsData.count === 1 ? 'it' : 'them'}.
```

loilo replied:
That's where I'm coming from, but it will [upset the linter](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md).

fregante replied:
’ isn't '

It shouldn't complain

fregante commented:
>  I could _not_ move the `isRepoWithAccess()` check outside the function

The reason is that you're using `nowAndOnAjaxedPages`; that only works if you use `elementReady` somewhere, otherwise it's run too soon. 

Since `isRepoWithAccess` basically just checks for the Settings tab, you could implement that manually, still outside the function, as `if(await elementReady('settings tab selector'))`

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,121 @@
+}, {
+	maxAge: 1 / 2, // Stale after 12 hours
+	staleWhileRevalidate: 2,
+	cacheKey: ([forkedRepo]): string => `${__featureName__}:pr-data:${forkedRepo ?? '-'}`
```
I get that we may dismiss `pr-data` as a case of premature optimization for possible future additions to the feature, but I think we should stick with the colon as a separator between feature name and remainder as that's how all other features are doing it.

fregante replied:
Oops, yeah, I forgot to add the `:`

Generally I'd even avoid using `__featureName__` and use the function's name (or intent) instead, like `'open-fork-prs:' + forkedRepo`

loilo replied:
Yep, generally liking that better as well. I'd be a little bit concerned about accidental unnoticed name clashes though...

loilo commented:
Sorry, I'm a little bit stumped on your last example. To do that ouside the function, we'd need top-level await (which we don't have, do we?)

Or are you suggesting to put it into a Promise?

```js
const hasSettings = elementReady('settings tab selector');
```

fregante commented:
It should go inside the `init` functions

loilo commented:
Okay, so that's basically how I had it in the first place. I ultimately outsourced all those checks into the `getOpenPullRequestsData` function purely as a matter of code deduplication (which is also why the function originally did not have any parameters). I can see how the function currently has too many unrelated responsibilities though.

Going to continue working on this tomorrow, it's getting late here and my coding as well as my writing is getting noticeably worse. 😁

loilo commented on source/features/show-open-prs-of-forks.tsx:
```diff
@@ -0,0 +1,121 @@
+			issueCount
+			nodes {
+				... on PullRequest {
+					url
```
What would be the incentive to assemble the URL programmatically?

fregante replied:
Smaller cache, we have 130+ features and we’re still limited to 5MB

loilo replied:
Ah, alright. So I guess storing an array would be preferrable to storing an object as well?

loilo commented:
Okay, I've renamed `getOpenPullRequestsData` to `getRawOpenPullRequestsData` which returns `[number, number] | false` to reduce the cache footprint. Init functions still call `getOpenPullRequestsData`, which is responsible for getting the raw (cached) data and assembles the object structure, including the URL.

fregante commented:
> I've renamed `getOpenPullRequestsData` to `getRawOpenPullRequestsData`

Good idea! I reorganized the code so "getOpenPullRequestsData" returns the full required data, which is `count` and `url`, so they can be used directly without further if/else

Can you resolve the conflicts?

fregante commented on source/libs/utils.ts:
```diff
@@ -88,6 +88,10 @@ export const getOwnerAndRepo = (): {
 	return {ownerName, repoName};
 };
 
+export function getForkedRepo(): string | undefined {
```
This is ambiguous. Is it the repo that was forked or the fork itself?

The correct term would be "source repo". Perhaps `getForkSourceRepo` would be the correct term.

Can you change this name everywhere? Including `const forkedRepo = ...`

loilo replied:
I thought so as well, but I trusted the existing naming (the function existed before, I just moved it to utils).

Yes, I'll take care of it when I'm going to handle #2934.

loilo replied:
Hm, there's already a function with that name in the `forked-to` feature.

```ts
const getForkSourceRepo = (): string => getForkedRepo() ?? getRepoURL();
```

fregante replied:
How about
- `getParentRepo`: only works if there’s a parent)
- `getSourceRepo`: always gets the source repo, whether it’s the parent or the current

loilo replied:
Sound's good to me. I'll do some refactoring of these, but I think I'll open a separate PR for that.

fregante commented on source/libs/utils.ts:
```diff
@@ -88,6 +88,10 @@ export const getOwnerAndRepo = (): {
 };
 
+export function getForkedRepo(): string | undefined {
+	return select<HTMLAnchorElement>('.fork-flag a')?.pathname.slice(1);
```
Also you could use the same meta tags I used in the `isForkedRepo` since they're available earlier on the page:

```
<meta name="octolytics-dimension-repository_parent_nwo" content="sindresorhus/refined-github" />
```

```suggestion
 	return select<HTMLMetaElement>('[name="octolytics-dimension-repository_parent_nwo"]')?.content;
```

loilo replied:
Sure. 👍
